### ISSUE TITLE
Holey array representation is incorrect in devtools

### ISSUE THREAD

**Issue by nchevobbe** [firefox-devtools/debugger#6143]
_From @nchevobbe on July 17, 2017 7:42_

_From [Bug 1380859](https://bugzilla.mozilla.org/show_bug.cgi?id=1380859)_

---

**Steps to reproduce:**

var x = new Array(10);
x.push(1);
console.log(x);

**Actual results:**

Array [ <10 empty slots>, more… ]

**Expected results:**

Array [ <10 empty slots>, 1 ]

_Copied from original issue: devtools-html/devtools-core#504_

**Comment by nchevobbe** [comment]
I think that the problem is that we don't get more than 10 preview items for arrays, and even if there are holes, they do count as preview items. Maybe we could change what the server sends in such cases, but it will be tricky to know where the holes are then.

**Comment by nchevobbe** [comment]
_From @Loirooriol on August 15, 2017 12:10_

The problem is the 10 preview limit indeed, e.g.

 - Input: `Object.assign([], {1e3: 1})` 
 - Result: `Array [ <10 empty slots>, … ]`
 - Expected: `Array [ <1000 empty slots>, 1 ]`

I attempted to fix this after I improved PseudoArray. My idea was allowing ArrayLike previews to be sparse maps, and iterate the previews using methods like `entries`, which belong to both arrays and maps. Each sequence of empty slots would count as 1 preview item of the maximum 10. There was still the problem of having empty slots at the end of the preview, I thought about adding the index of the last empty slot after the last entry in the preview, or including the next non-empty entry even if it wouldn't end up displayed. It was tricky indeed and I abandoned the patch.
